### PR TITLE
Add HLRC for EQL search

### PR DESCRIPTION
**Pull request by imotov**
Adds EQL HLRC client with the search method.

Relates to #51961

This PR will add just search. We will add stats and translate in follow up iterations.

### CODE REVIEW

**Review by costin — APPROVED** [review]
LGTM

**Review by aleksmaus — APPROVED** [review]
👍 
Thank you!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchRequest.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    public int fetchSize() {
+        return this.fetchSize;
+    }
+
+    public EqlSearchRequest fetchSize(int size) {
+        this.fetchSize = size;
+        if (fetchSize <= 0) {
+            throw new IllegalArgumentException("size must be more than 0");
```
Comment (astefan) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/eql/EqlSearchRequest.java]: Maybe better as `size must be greater than 0`?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchRequestTests.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.List;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class EqlSearchRequestTests extends AbstractRequestTestCase<EqlSearchRequest, org.elasticsearch.xpack.eql.action.EqlSearchRequest> {
+
+    @Override
+    protected EqlSearchRequest createClientTestInstance() {
+        EqlSearchRequest EqlSearchRequest =
+            new EqlSearchRequest("testindex", randomAlphaOfLength(40));
```
Comment (astefan) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchRequestTests.java]: Maybe use one line for these?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchRequestTests.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        if (randomBoolean()) {
+            EqlSearchRequest.fetchSize(randomIntBetween(0, Integer.MAX_VALUE));
```
Comment (astefan) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchRequestTests.java]: fetchSize is required to be strictly positive in `EqlSearchRequest`.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchRequestTests.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    protected void assertInstances(org.elasticsearch.xpack.eql.action.EqlSearchRequest serverInstance, EqlSearchRequest
+        clientTestInstance) {
+        assertThat(serverInstance.eventTypeField(), equalTo(clientTestInstance.eventTypeField()));
+        assertThat(serverInstance.implicitJoinKeyField(), equalTo(clientTestInstance.implicitJoinKeyField()));
+        assertThat(serverInstance.timestampField(), equalTo(clientTestInstance.timestampField()));
+        assertThat(serverInstance.query(), equalTo(clientTestInstance.query()));
+        assertThat(serverInstance.rule(), equalTo(clientTestInstance.rule()));
+        assertThat(serverInstance.searchAfter(), equalTo(clientTestInstance.searchAfter()));
+        assertThat(serverInstance.indicesOptions(), equalTo(clientTestInstance.indicesOptions()));
+        assertThat(serverInstance.indices(), equalTo(clientTestInstance.indices()));
```
Comment (astefan) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchRequestTests.java]: I think `fetchSize` is missing from these checks.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchResponseTests.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        if (randomBoolean()) {
+            hits = new ArrayList<>();
+            for (int i = 0; i < size; i++) {
+                hits.add(new SearchHit(i, randomAlphaOfLength(10), new HashMap<>()));
+            }
+        }
+        if (randomBoolean()) {
+            return hits;
+        }
+        return null;
```
Comment (astefan) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/eql/EqlSearchResponseTests.java]: To me it looks more common to return `hits` and less common to return `null` and I would have switched between them. But that's a personal preference.

**Review by astefan — COMMENTED** [review]
Left few minor comments.

File: client/rest-high-level/build.gradle [hunk 6]
```diff
@@ -128,6 +130,10 @@ testClusters.all {
   setting 'xpack.security.authc.api_key.enabled', 'true'
   setting 'xpack.security.http.ssl.enabled', 'false'
   setting 'xpack.security.transport.ssl.enabled', 'false'
+  if (BuildParams.isSnapshotBuild() == false) {
+    systemProperty 'es.eql_feature_flag_registered', 'true'
```
[the remark this answers is not in the document]
Comment (imotov) [reply client/rest-high-level/build.gradle]: I am following the same pattern as CCR here. The setting exist only in the snapshot build unless we have this environment variable set, then it shows up in the release build as well. This way we test it in both release and snapshot builds only through slightly different mechanisms.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by imotov** [comment]
Thank you for the review everyone!

**Comment by nik9000** [comment]
This broke my ability to run `x-pack:plugin:analytics` tests in my IDE because `analytics` pulls in the hlrc for testing and the hlrc pulls in *this* for testing. Somehow that causes two versions of guava to be on the classpath. Could we flip the testing dependency so this depends on hlrc instead of hlrc depending on it? That way testing the Elastic licensed bits depends on the Apache licensed bits. And it makes Eclipse happy.

**Comment by costin** [comment]
@nik9000 Please raise an issue - this comment is valuable but I think most folks haven't seen it considering the PR has been merged.

**Pull request merged**